Ports/build_all.sh

```diff
@@ -1,89 +1,109 @@
-            fi
+log_success() {
+    echo -e "\033[1;32m[#]\033[0m $1"
+}
+log_warn() {
+    echo -e "\033[1;33m[!]\033[0m $1"
+}
+log_info() {
+    echo -e "\033[1;36m[*]\033[0m $1"
+}
+log_process() {
+    echo -e "\033[1m[~]\033[0m $1"
+}
+log_error() {
+    echo -e "\033[1;31m[x]\033[0m $1"
+}
+
+do_build_port() {
+    log_process "Building $port_name"
+    if $verbose; then
+        ./package.sh
+    else
+        ./package.sh &> /dev/null
+    fi
+}
 
-            if [ "$clean" == true ]; then
-                if [ "$verbose" == true ]; then
-                    ./package.sh clean_all
-                else
-                    ./package.sh clean_all > /dev/null 2>&1
+do_clean_port() {
+    log_process "Cleaning $port_name"
+    if $verbose; then
+        ./package.sh "$1"
+    else
+        ./package.sh "$1" &> /dev/null
+    fi
+}
+
+# It is required to simply implement cd, instead of pushd-popd
+ports_dir=$(realpath "$(dirname "${BASH_SOURCE[0]}")")
+for port_dir in $(find "$ports_dir" -mindepth 1 -maxdepth 1 -type d | sort); do
+    port_name="$(basename "$port_dir")"
+    if [[ " ${processed_ports[*]} " == *" $port_name "* ]]; then
+        log_info "$port_name is already processed"
+        continue
+    fi
+
+    if ! cd "$port_dir"; then
+        log_error "Can not change directory to '$port_name'"
+        exit 1
+    fi
+
+    if [[ ! -x ./package.sh ]]; then
+        log_warn "$port_name does not have executable package.sh"
+        continue
+    fi
+
+    case "$action" in
+        clean*)            
+            if do_clean_port "$action"; then
+                log_success "Cleaned $port_name"
+            else
+                log_error "Failed cleaning $port_name"
+                if $failfast; then
+                    exit 1
                 fi
             fi
-            if [ "$verbose" == true ]; then
-                if ./package.sh; then
-                    echo "Built ${port}."
-                else
-                    echo "ERROR: Build of ${port} was not successful!"
-                    some_failed=true
-                    popd > /dev/null
-                    continue
-                fi
+        ;;
+        build)
+            if do_build_port; then
+                log_success "Built $port_name"
             else
-                if ./package.sh > /dev/null 2>&1; then
-                    echo "Built ${port}."
-                else
-                    echo "ERROR: Build of ${port} was not successful!"
-                    some_failed=true
-                    popd > /dev/null
-                    continue
+                log_error "Failed building $port_name"
+                some_failed=true
+                if $failfast; then
+                    exit 1
                 fi
             fi
+        ;;
+    esac
 
-            built_ports="$built_ports $port $(./package.sh showproperty depends) "
-        popd > /dev/null
-    fi
+    processed_ports+=("$port_name" "$(./package.sh showproperty depends)")
```

Comment:
With the current change its catching all the ports. I tried `echo ${processed_ports[@]`

```
aclock ncurses acpica-tools  alpine openssl ncurses ca-certificates angband ncurses SDL2 SDL2_image SDL2_ttf SDL2_mixer Another-World SDL2 zlib aria2 libssh2 libxml2 openssl zlib libuv awk  backward-cpp binutils zlib bash  bass scummvm bc  bdwgc libatomic_ops bison  bochs SDL2 boost zlib bzip2 zstd xz libicu brogue SDL2 SDL2_image brotli  byacc  bzip3  c-ares  carl  cavestory libjpeg libpng SDL2 SDL2_image SDL2_mixer SDL2_ttf cbonsai ncurses ccache zstd cfunge  chester SDL2 chocolate-doom libpng libsamplerate SDL2 SDL2_mixer SDL2_net citron sparsehash libffi pcre ClassiCube SDL2 curl libopenal cmake bash make sed ncurses libuv openssl cmatrix ncurses composer php coreutils  cowsay perl5 cpio  c-ray SDL2 dash  deutex libpng dialog ncurses diffutils libiconv dmidecode  doom SDL2 SDL2_mixer dos2unix gettext dosbox-staging libslirp libpng opusfile SDL2 SDL2_net dosfstools  double-conversion  drascula scummvm dreamweb scummvm dropbear  dtc bash dungeonrush SDL2 SDL2_image SDL2_mixer SDL2_ttf SDL2_net e2fsprogs  ed pcre2 edid-decode  editline  emu2  epsilon SDL2 libpng libjpeg freetype expat  ffmpeg libiconv libtiff xz bzip2 SDL2 x264 x265 figlet  file  findutils  fio zlib flac libogg flatbuffers  flex m4 pcre2 fontconfig libxml2 freetype fotaq scummvm freeciv SDL2 SDL2_image SDL2_mixer SDL2_ttf SDL2_gfx zstd libicu xz gettext curl freedink SDL2 SDL2_image SDL2_mixer SDL2_ttf SDL2_gfx gettext fontconfig glm frotz ncurses gawk  gcc binutils gmp mpfr mpc isl gdb gmp binutils gemrb freetype libiconv python3 SDL2 SDL2_mixer zlib genemu SDL2 genext2fs  git zlib curl glib gettext libffi libiconv pcre zlib gltron libpng glu SDL_sound sdl12-compat zlib gn ninja python3 gnuapl  gnucobol gmp gcc bash ncurses gnupg libiconv libgpg-error libgcrypt libksba libassuan npth ntbtls gnuplot  gperf  grep  grepcidr  griffon scummvm gsl  guile gmp libunistring libffi bdwgc libiconv gzip  halflife xash3d-fwgs harfbuzz freetype libicu hatari SDL2 zlib hexedit ncurses imagemagick libpng libtiff libjpeg imgcat ncurses libpng libjpeg termcap indent  ja2 SDL2 jakt llvm jdupes libjodycode jfduke3d SDL2 joe  jot libpuffy jq  kakoune bash sed klong  less ncurses libarchive zlib pcre libexpat  libfftw3f  libgd libpng libmad  libmikmod  libmodplug  libmpeg2  libmpg123  liboggz libogg libopus  libphysfs  libsixel  libsodium  libtheora libvorbis libtool bash sed libuuid  libyaml  libzip zlib links  lite-xl freetype lua pcre2 SDL2 lowdown  lrzip bzip2 lz4 lzo zlib luajit  luarocks lua git readline lure scummvm lynx ncurses openssl zlib lzop lzo mandoc less pcre2 zlib mawk  mbedtls  mc gettext glib libtool ncurses vim md4c  mednafen SDL2 zlib flac mgba SDL2 zlib sqlite libpng libzip milkytracker SDL2 zlib mold zlib openssl zstd mrsh  mruby  mysthous scummvm nano ncurses nasm  ncdu ncurses neofetch bash jq nesalizer SDL2 nethack ncurses bash nippon scummvm nlohmann-json  npiet libgd libpng nyancat  ObjFW openssl oksh ncurses OpenJDK fontconfig libffi openrct2 curl flac fontconfig freetype libicu libogg libpng libvorbis libzip nlohmann-json openssl SDL2 speexdsp zlib openssh zlib openssl openttd freetype SDL2 libicu libpng zlib xz openttd-opengfx openttd-opensfx opentyrian SDL2 SDL2_net opfor xash3d-fwgs optipng  p7zip libiconv pacman SDL2 SDL2_image SDL2_ttf SDL2_mixer patch  pfetch bash pixman  pkgconf  poppler fontconfig freetype libjpeg libpng libtiff potrace zlib powdertoy luajit curl libfftw3f zlib SDL2 prboom-plus glu libmad libvorbis SDL2 SDL2_image SDL2_mixer SDL2_net printf libpuffy pt2-clone SDL2 pv  qemu glib libslirp pixman SDL2 qoi libpng stb qt6-qt5compat qt6-qtbase libiconv qt6-serenity qt6-qtbase quake SDL2 quake2 SDL2 quake3 SDL2 r0  radare2  RetroArch freetype SDL2 zlib RISCVEmu  rsync  ruby  rvvm sdl12-compat sam SDL2 SDL2-GNUBoy SDL2 SDL2_sound SDL2 SDL_mixer libmikmod libvorbis sdl12-compat timidity SDLPoP SDL2 SDL2_image serenity-theming  serious-sam-classic libvorbis SDL2 zlib sfinx scummvm sl ncurses soltys scummvm ssmtp openssl stockfish  stpuzzles  stress-ng zlib Super-Mario SDL2 SDL2_mixer SDL2_image SuperTuxKart curl freetype glu harfbuzz libjpeg libopenal libogg libpng libvorbis openssl SDL2 sqlite zlib tar  taskwarrior libuuid tcl  thesilversearcher pcre xz tig ncurses pcre readline libiconv tinycc  tinyscheme  tr libpuffy tree  tuxracer glu SDL_mixer sdl12-compat tcl vitetris  vlang  vttest  VVVVVV SDL2 SDL2_mixer wayland expat libxml2 wget openssl which  wireguard-tools  yasm  zig  zsh
```

Please confirm, should it be changed to `processed_ports+=("$port_name" $(./package.sh showproperty depends))`?

Reply:
> Please confirm, should it be changed to processed_ports+=("$port_name" $(./package.sh showproperty depends))?

Yes, I believe so. Right now it doesn't make a difference, because command substitution strips trailing newlines and we combine everything into a single string anyways, but if someone else wants to use that array for something and doesn't know this it might be a problem.

The difference between both is obvious if one uses `for i in "${processed_ports[@]}"; do echo "${i}"; done` to print the array instead.

With quotes around the command substitution:

```
[#] Built angband
aclock
ncurses
acpica-tools

alpine
openssl ncurses ca-certificates
angband
ncurses SDL2 SDL2_image SDL2_ttf SDL2_mixer
```

Without quotes around the command substitution:

```
[#] Built angband
aclock
ncurses
acpica-tools
alpine
openssl
ncurses
ca-certificates
angband
ncurses
SDL2
SDL2_image
SDL2_ttf
SDL2_mixer
```

Reply:
Fixed!